### PR TITLE
Feat/swarinfo 364 state tracer

### PR DESCRIPTION
**Pull request by cquesnel**
Added a way to debug state transitions by logging the transitions as well a a transition event in a queue.

### CODE REVIEW

File: src/bsp/src/stm32/src/interloc/src/InterlocManager.cpp [hunk 1]
```diff
@@ -41,8 +41,8 @@ void InterlocManager::startInterloc() {
     syncClocks();
 
     // Uncomment one of the following lines to go into TWR
-    // m_stateHandler.setState(InterlocStates::WAIT_POLL);
-    // m_stateHandler.setState(InterlocStates::SEND_POLL);
+    // m_stateHandler.setState(InterlocStates::WAIT_POLL, InterlocEvent::NO_EVENT);
+    // m_stateHandler.setState(InterlocStates::SEND_POLL, InterlocEvent::NO_EVENT);
```
Comment (stlaperriere) [inline src/bsp/src/stm32/src/interloc/src/InterlocManager.cpp:45]: Commented code?

File: src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp [hunk 2]
```diff
@@ -2,9 +2,22 @@
...
-    m_currentState(&InterlocStateContainer::getState(InterlocStates::DEFAULT)) {}
+    m_currentStateName(InterlocStates::DEFAULT),
+    m_currentState(&InterlocStateContainer::getState(InterlocStates::DEFAULT)),
+    m_stateTracer(m_stateTracerData.data(), MAX_TRACER_TRANSITIONS) {}
 
-void InterlocStateHandler::setState(InterlocStates state) {
+void InterlocStateHandler::setState(InterlocStates state, InterlocEvent event) {
+    if (m_stateTracer.isFull()) {
+        m_stateTracer.pop();
+    }
```
Comment (stlaperriere) [inline src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp:12]: Why the need to pop if m_stateTracer is a circular queue?

File: src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp [hunk 3]
```diff
@@ -2,9 +2,22 @@
...
+
+    auto& stateTransition = m_stateTracer.getNextAllocation().value().get();
+    stateTransition.m_fromState = m_currentStateName;
+    stateTransition.m_toState = state;
+    stateTransition.m_transitionEvent = event;
+    m_stateTracer.push(stateTransition);
```
Comment (stlaperriere) [inline src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp:18]: You push the same value you just got from the queue? Won't this just invariably overwrite the last value?

**Review by stlaperriere — CHANGES_REQUESTED** [review]
I have questions.

File: src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp (hunk 2 above)
Comment (cquesnel) [reply src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp:12]: The queue won't push if it's full. So we clear the oldest data before pushing

File: src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp (hunk 3 above)
Comment (cquesnel) [reply src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp:18]: The getNextAllocation() returns a reference to the next empty slot in the queue, so I change the data in that slot. The push then just changes the counters in the queue. The push uses a reference and doesn't copy the data, so if I didn't use the getNextAllocation() first, the data I push would go out of scope the moment I exit the function.

File: src/bsp/src/stm32/src/interloc/src/InterlocManager.cpp (hunk 1 above)
Comment (cquesnel) [reply src/bsp/src/stm32/src/interloc/src/InterlocManager.cpp:45]: Yes, this starts the interloc in one of two modes, but won't work if no hivesight is on the board so it is just there to make it easier when testing without having to always rewrite these lines

File: src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp (hunk 3 above)
Comment (stlaperriere) [reply src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp:18]: Oh okay I see.

**Review by stlaperriere — APPROVED** [review]

File: src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp (hunk 3 above)
Comment (xgroleau) [reply src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp:18]: The push is supposed to copy the data though, as of now you are copying twice and not calling the destructor when overwriting it (which is not an issue in this case)

**Review by xgroleau — CHANGES_REQUESTED** [review]

File: src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp (hunk 3 above)
Comment (xgroleau) [reply src/bsp/src/stm32/src/interloc/src/InterlocStateHandler.cpp:18]: getNextAllocation should only be used with advance afterwards. push == getNextAllocation + advance. It's just if you want the  variable temporarly and the confirm if you advance (or not), you can use getNextAllocation

**Review by xgroleau — APPROVED** [review]

**Pull request merged**